### PR TITLE
feat(webpack): add compilation dependency sets

### PR DESCRIPTION
**Pull request by emilio-martinez**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: (See below)
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---

- Added `Compilation` dependency properties. See https://github.com/webpack/webpack/blob/v4.16.3/lib/Compilation.js#L2150.
- These properties are of a SortableSet type, which is defined at https://github.com/webpack/webpack/blob/v4.16.3/lib/util/SortableSet.js. Interface added to describe that type.

### CODE REVIEW

File: types/webpack/index.d.ts
Comment (alan-agius4) [inline types/webpack/index.d.ts]: I don't think this is totally correct, unless I am missing something.

`SortableSet` is a interface, but the a set is a `Class`, thus if you try to;

```ts
const x = new SortableSet([]);
```

You will end up with an error `SortableSet' only refers to a type, but is being used as a value here.`

Comment (emilio-martinez) [reply types/webpack/index.d.ts]: I agree with the class vs interface. That said, I guess that brings me to a question around how to treat private APIs in type defs. Webpack doesn't export that util at their top-level (see https://github.com/webpack/webpack/blob/HEAD/lib/webpack.js#L165); in other words, I'm not sure if users are intended to do `new SortableSet([])` at all. That's why I made it an interface vs a class. To import `SortableSet` you'd have to do something like `import {} from 'webpack/lib/util/SortableSet`.

Happy to change to a class—it's a quick change. However, I'm unclear on how to handle due to the above. I could even put `SortableSet` in a `util` namespace, but ultimately you wouldn't actually be importing the class in Javascript. Let me know your thoughts.

**Comment by johnnyreilly** [comment]
Is this safe to merge?

File: types/webpack/index.d.ts
Comment (alan-agius4) [reply types/webpack/index.d.ts]: This cannot be a private since it’s being used by an exported member.

If the sortableset is not meant to be used as a class, I don’t have any problem with the current implementation using an interface.

That said, I am gonna loop in @Hotell, for another option.

**Review by alan-agius4 — APPROVED** [review]
After thinking about it, I approve, Unless anyone has a better idea on how to get around this.

This is a good fix, and if needs be,
we can eventually improve this and it shouldn’t cause breaking changes.

Ps: thanks for your contribution.

File: types/webpack/index.d.ts
Comment (Hotell) [reply types/webpack/index.d.ts]: yo folks! sorry to be kinda late to the party. If `SortableSet` is a runtime entity ( which is -> class ) it should be typed as runtime executable , in our case a es2015 class. Also it should be moved under proper path like within implementation, so consumer wont be able to do `const {SortableSet} = require('webpack')`

**Pull request merged**
